### PR TITLE
Fix proactive cancel.

### PR DESCRIPTION
**Pull request by boaks**
Postpone proactive cancel, if reregistration is already pending.

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by sbernard31** [comment]
(I will look at this tomorrow)

**Comment by boaks** [comment]
> By the way, is it possible that if we face a race condition too if we remove/cancel the observation from store at the same time we do a reregister ? (it would concern proactive or reactive cancel)

I don't get that. FMPOV, we should always, ASAP mark the observation as canceled. The inhibits the most reregistrations. Just in the case a reregistration is already pending, the cancel requires special processing.

**Comment by boaks** [comment]
> In the case of the reregister/proactiveCancel race condition, it makes sense to me to not send cancel and reregister at same time but maybe this is not needed to sendproactiveCancel onResponse as RST must be send for unwanted notify ?

I would assume, that using the proactive cancel is intended for cases, where for some reasons the coap-server should be informed about that cancel ASAP. It's not sure, that the reregistration request get's rejected (that's also a race-condition at the end or the reregistration), so that approach would require additional checks, to ensure the rejection. Therefore I stick to send that "cancel-request". In my opinion (test experience) that all is really rare, so I don't think spending the time in optimization pays off.

File: californium-core/src/main/java/org/eclipse/californium/core/CoapObserveRelation.java [hunk 1]
```diff
@@ -71,12 +71,12 @@
 	private final long reregistrationBackoff;
 
 	/**
-	 * Indicates, that an observe request is pending.
+	 * Indicates, that an observe request or cancel observe request is pending.
 	 * 
 	 * {@ink #reregister()} is only effective, when no observe request is
```
[the remark this answers is not in the document]
Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/CoapObserveRelation.java]: Fixed

File: californium-core/src/main/java/org/eclipse/californium/core/CoapObserveRelation.java [hunk 2]
```diff
@@ -86,6 +86,8 @@
 
 	/** Indicates whether the relation has been canceled. */
 	private volatile boolean canceled = false;
+	/** Indicates whether the relation has been canceled. */
```
[the remark this answers is not in the document]
Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/CoapObserveRelation.java]: fixed

**Comment by boaks** [comment]
Some more thoughts on RST instead of cancel-request:
If the the reregister request gets a piggybacked ACK as response, RST will be no option.

**Comment by sbernard31** [comment]
Agree, but RST will be sent for next notify ?

**Comment by boaks** [comment]
Yes, but that would then result in not inform the server ASAP and so breaks the contract of the proactiveCancel. (By the way: I don't think, proactive cancel will be used too often, because in many cases it's quite good enough to reject future notification.)

**Pull request merged**